feat: add switch

tanvibhakta opened the pull request:
## Description

Fixes https://github.com/appsmithorg/appsmith/issues/19321

Zeplin: https://zpl.io/Nwd7B59

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
https://github.com/appsmithorg/TestSmith/issues/2141

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

albinAppsmith commented on packages/design-system-v2/src/Toggle/Toggle.stories.tsx:
```diff
@@ -0,0 +1,11 @@
+  component: Toggle,
+} as ComponentMeta<typeof Toggle>;
+
+export const Basic = <Toggle />;
```
Can we make this not nested? I don't see a variation for toggle in future.

albinAppsmith commented on packages/design-system-v2/src/Toggle/Toggle.styles.tsx:
```diff
@@ -0,0 +1,57 @@
+
+  &:hover {
+    // TODO: needs a theme variable
+    --toggle-color-bg: var(--ads-v2-color-black-250)
```
we do have theme variable, --ads-v2-color-bg-emphasis

tanvibhakta replied:
that is the background color already. We need a theme variable for black-250.

albinAppsmith commented on packages/design-system-v2/src/Toggle/Toggle.styles.tsx:
```diff
@@ -0,0 +1,57 @@
+    // set position of thumb when checked
+    &::before {
+      grid-area: on;
+    }
```
:hover state for on state is also needed.

tanvibhakta replied:
We apparently only have hover for default state, not active. https://www.figma.com/file/lWkmRog0cbht9upiFJVs4s?node-id=15:4280#347892275

albinAppsmith commented on packages/design-system-v2/src/Toggle/Toggle.styles.tsx:
```diff
@@ -0,0 +1,57 @@
+
+  &:disabled {
+    cursor: not-allowed;
+    opacity: 0.5;
```
opacity is 0.6, may be we can add it in theme

albinAppsmith commented on packages/design-system-v2/src/Toggle/Toggle.tsx:
```diff
@@ -0,0 +1,23 @@
+import { ToggleProps } from "./Toggle.types";
+
+const Toggle = (props: ToggleProps) => {
+  const { checked, ...rest } = props;
```
onChange needs to be destructured, else we will face issue with main repo

albinAppsmith commented on .changeset/moody-files-enjoy.md:
```diff
@@ -0,0 +1,6 @@
+---
+"design-system-v2": patch
```
```suggestion
```

albinAppsmith commented on packages/design-system/src/__theme__/default/index.css:
```diff
@@ -1,6 +1,10 @@
 @import "colors.css";
 @import "variables.css";
 
+*:focus-visible {
```
Remove this

albinAppsmith commented on packages/design-system/src/__theme__/default/index.css:
```diff
@@ -119,6 +121,8 @@
     * ===========================================*
   */
   --ads-v2-border-radius: 4px;
+  --ads-v2-pill-border-radius: 10px;
```
Do you see any other use case for this variable? other than switch?

tanvibhakta replied:
not particularly but wasn't sure about hardcoding it - shall I?

albinAppsmith replied:
Yeah seems like this is only foe switch component. So keep it inside that component only

albinAppsmith commented on packages/design-system/src/Switch/Switch.styles.tsx:
```diff
@@ -0,0 +1,86 @@
+export const StyledSwitchInput = styled.input<{
+  isDisabled?: boolean;
+  isSelected?: boolean;
+  onClick?: any;
```
Why onClick type declared here?

tanvibhakta replied:
Will remove.

shadabbuchh commented:
Bugs in this PR:

**Functional:**

1.  When we enter a long text in the `children` control, the switch component is not completely visible.
<img width="1184" alt="Screenshot 2023-02-06 at 8 13 01 PM" src="https://user-images.githubusercontent.com/39921438/217002051-61615d4d-d6e7-4d97-a762-2ac299bc7da5.png">

2. Upon selecting a value for `isSelected`, we cannot change the selection from the canvas.
3. In disabled state we get a hover state
https://www.loom.com/share/50af0b9228cb4f9dbbe799d4db5d7f5a

**Design:**

1. Line height for the Label is 19px whereas the expected value is 20 px.
2. Letter spacing is -0.24px, expected is -0.1px.
3. Wrong background colour for unselected -> hover state.
5. Not able to see Border colour in Focus state.
6. Focus state looks different in different browsers: https://www.loom.com/share/f57abf49a3ca48a88305de57b8dc16fc

tanvibhakta commented:
Functional:

>    When we enter a long text in the children control, the switch component is not completely visible.
Fixed.

>    Upon selecting a value for isSelected, we cannot change the selection from the canvas.
This happens across components for a "controlled" state. Controlled means that the programmer now needs to implement the control if they want the value of it. 

>    In disabled state we get a hover state
This was happening only when the switch was selected and then disabled. Fixed.


Design:

>    Line height for the Label is 19px whereas the expected value is 20 px.
>    Letter spacing is -0.24px, expected is -0.1px.
We're not looking at typography here, right? 

>    Wrong background colour for unselected -> hover state.
fixed

>    Not able to see Border colour in Focus state.
>    Focus state looks different in different browsers: https://www.loom.com/share/f57abf49a3ca48a88305de57b8dc16fc
both should be fixed now.

shadabbuchh commented:
All of the above issues are fixed.
Additional issues noticed:

1. No hover state when the toggle is selected
2. Wrong Box shadow in all states when the toggle is selected.
@tanvibhakta

shadabbuchh commented:
All issues are fixed. PR is good to go.
